lib/ECDSA.ts

```diff
@@ -0,0 +1,121 @@
+import * as Elliptic from 'elliptic';
+import * as lapo_asn1js from '@lapo/asn1js';
+import { unarmor } from '@lapo/asn1js/base64';
+
+export enum CurveName { secp256k1 = 'secp256k1' }
+
+export interface ECDSAPublicKey {
+  tag: 'ECDSAPublicKey';
+  pem: string;
+  curveName: CurveName;
+}
+
+// not used in CSL but for API purposes we create a proper datatype for it.
+// This is also why there is no PrivateKeyValue
+export interface ECDSAPrivateKey {
+  tag: 'ECDSAPrivateKey';
+  pem: string;
+  curveName: CurveName;
+}
+
+export const mkECDSAPrivateKey = (pem: string): ECDSAPrivateKey => ({
+  pem,
+  tag: 'ECDSAPrivateKey',
+  curveName: CurveName.secp256k1,
+});
+
+export interface ECDSASignature {
+  tag: 'ECDSASignature';
+  signature: {
+    bytes: string,
+  };
+}
+
+const globalEC = new Elliptic.ec(CurveName.secp256k1);
+
+/**
+ * @returns a binary buffer or a string failure
+ */
+function decodeASN1PublicPem(pem: string): Buffer | string {
+  /*
+  Informal ASN.1 Schema of openssl elliptic curve private keys (undocumented).
+  You can find it (or something like it) by running `openssl asn1parse -in [PUBLIC_KEY] -dump`
+  OpenSSLECPublicKey ::= SEQUENCE {
+    metadata    SEQUENCE {
+      publicKeyType     OBJECT IDENTIFIER 1.2.840.10045.2.1,
+      ellipticCurve     OBJECT IDENTIFIER 1.2.840.10045.3.1.7,
+    }
+    publicKey   BIT STRING
+  }
+  */
+  try {
+    const decoded = lapo_asn1js.decode(unarmor(pem));
+    if (!decoded.sub || !decoded.sub[1]) {
+      return `Unexpected ASN.1 decode result:\n ${decoded}`;
+    }
+    const publicKey = decoded.sub[1];
+
+    // hexdump the bitstring of the public key
+    const bits =
+      publicKey.stream.hexDump(publicKey.posContent(), publicKey.posEnd(), true).substring(2);
+    return Buffer.from(bits, 'hex');
+  } catch (err) {
+    return err.toString();
+  }
+}
+
+/**
+ * @returns a binary buffer or a string failure
+ */
+function decodeASN1PrivatePem(pem: string): Buffer | string {
+  /*
+  ASN.1 Schema of openssl elliptic curve private keys (https://tools.ietf.org/html/rfc5915):
+    ECPrivateKey ::= SEQUENCE {
+      version        INTEGER { ecPrivkeyVer1(1) } (ecPrivkeyVer1),
+      privateKey     OCTET STRING,
+      parameters [0] ECParameters {{ NamedCurve }} OPTIONAL,
+      publicKey  [1] BIT STRING OPTIONAL
+    }
+  */
+  try {
+    const decoded = lapo_asn1js.decode(unarmor(pem));
+    if (!decoded.sub || !decoded.sub[1]) {
+      return `Unexpected ASN.1 decode result:\n ${decoded}`;
+    }
+    const privateKey = decoded.sub[1];
+    const hexstr = privateKey.stream.hexDump(privateKey.posContent(), privateKey.posEnd(), true);
+    return Buffer.from(hexstr, 'hex');
+  } catch (err) {
+    return err.toString();
+  }
+}
+
+export function decodePublicKey(pem: string, ec = globalEC): Elliptic.ec.KeyPair | string {
```

Comment:
Why does `decodePublicKey` and `decodePrivateKey` take an optional `Elliptic.ec`? It does not seem like it is passed in anywhere

Reply:
Yeah, it's because I originally didn't have the global elliptic curve generator, so I passed it in to functions. In general, we only support one type of curve right now which is secp256k1 but we might support others in the future, so I felt leaving in this option was prudent, but perhaps it's a YAGNI instance.